### PR TITLE
Asura: attempt to find new url

### LINKED ISSUE #16251: Asura Scans: HTTP error 404 and Too Many Follow-up Requests
### Source information

Asura Scans 1.4.45

### Source language

All language 

### Steps to reproduce

1. Open the chapter list of an old ongoing series.
2. Refresh the chapter list.
3. HTTP error 404

### Expected behavior

Chapter Refreshed 

### Actual behavior

Error 404

### Tachiyomi version

0.14.6

### Android version

Android 12

### Other details

I found out that the link from tachiyomi chapter list (chapter list web view) and the actual link (Asura Scans site homepage -> chapter list of the series) are different.

Example
Link from tachiyomi chapter list : https://asura.gg/manga/player-who-returned-10000-years-later/

The actual link : https://www.asurascans.com/manga/8239705535-player-who-returned-10000-years-later/

*Might be redirecting issues.
** Doesn't effect new series.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.14.6](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I have tried the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting/).
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

### LINKED ISSUE #16339: Asura Scans (Turkish): webtoon pages not found
### Source information

Asura Scans (Turkish): webtoon pages not found"

Asura Scans 1.4.45 

### Source language

Turkish 

### Steps to reproduce

Asura Scans turkish webtoon or manga appear, new episodes are added, but webtoons manga is not reading. always and On all webtoons say "page not found". Other sites or webtoons are running. I moved Webtoons but still the asura is broken. I did everything but stilll broke. AND This is just broken for TURKİSH ASURA. PLEASE HELP US
(I renewed the extensions but they are still broken) 

### Expected behavior

The problem must be solved and not live again. 

### Actual behavior

MY WEBTOONS MUST BE SAVED

### Tachiyomi version

0.14.6

### Android version

Android 

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.14.6](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I have tried the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting/).
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/tachiyomiorg/tachiyomi/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by mobi2002**
close #16251
close #16339

also fixes no page on Turkish version of extension

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt [hunk 1]
```diff
@@ -29,31 +24,87 @@ open class AsuraScans(
...
+                response.close()
+
+                val manga = getNewManga(url)
+                    ?: throw Exception("Migrate from Asura to Asura")
+
+                val newRequest = when (frag) {
+                    "chapters" -> chapterListRequest(manga)
+                    "details" -> mangaDetailsRequest(manga)
+                    else -> throw IllegalStateException("unknown url fragment for urlChangeInterceptor")
+                }
```
Comment (stevenyomi) [inline multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt:52]: Throw `IOException`s in the interceptor.

File: multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt [hunk 2]
```diff
@@ -70,55 +153,32 @@ open class AsuraScans(
...
-            summary = PREF_PERM_MANGA_URL_SUMMARY
-            setDefaultValue(true)
-
-            setOnPreferenceChangeListener { _, newValue ->
-                val checkValue = newValue as Boolean
-                preferences.edit()
-                    .putBoolean(getPermanentMangaUrlPreferenceKey(), checkValue)
-                    .commit()
-            }
+    // builtin `.putIfAbsent` requires min api 24
```
Comment (stevenyomi) [inline multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt:172]: You can use `MutableMap.getOrPut()` and throw away the result.

File: multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt [hunk 3]
```diff
@@ -29,31 +31,108 @@ open class AsuraScans(
...
+                "chapters" -> chapterListRequest(manga)
+                "details" -> mangaDetailsRequest(manga)
+                else -> throw IOException("unknown url fragment for urlChangeInterceptor")
+            }
+        }
     }
 
     override val client: OkHttpClient = network.cloudflareClient.newBuilder()
         .addInterceptor(uaIntercept)
+        .addInterceptor(urlChangeInterceptor)
```
[the remark this answers is not in the document]
Comment (stevenyomi) [reply multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt:112]: That is not possible. That interceptor attempts to make a second request, which a network interceptor is not capable of.

File: multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt [hunk 4]
```diff
@@ -29,31 +31,108 @@ open class AsuraScans(
...
+    override fun fetchMangaDetails(manga: SManga): Observable<SManga> {
+        val newManga = manga.apply {
+            url = "$url#details#${title.toSearchQuery()}"
+        }
 
-    override fun fetchSearchManga(page: Int, query: String, filters: FilterList): Observable<MangasPage> {
-        return super.fetchSearchManga(page, query, filters).tempUrlToPermIfNeeded()
+        return if (manga.initialized) {
+            // to mitigate extra requests caused by urlChangeInterceptor
+            Observable.just(manga)
```
[the remark this answers is not in the document]
Comment (mobi2002) [reply multisrc/overrides/mangathemesia/asurascans/src/AsuraScans.kt:132]: When the user pulls to refresh or library update with update details setting, the app makes manga details and chapter list requests together. correct me if I am wrong tho. This means both would make separate search requests as the slugMap isn't yet updated in the preference.

I can revert this, I added this when slugMap was not cached.

**Pull request merged**